profiling/profiler.go

```diff
@@ -0,0 +1,52 @@
+package profiling
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	hpprof "net/http/pprof"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// Profiler runs and controls a HTTP pprof interface.
+type Profiler struct {
+	mu     sync.Mutex
+	server *http.Server
+}
+
+// NewProfiler creates an instance of the profiler with
+// the given port.
+func NewProfiler(port int) *Profiler {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/debug/pprof/", hpprof.Index)
+	mux.HandleFunc("/debug/pprof/cmdline", hpprof.Cmdline)
+	mux.HandleFunc("/debug/pprof/profile", hpprof.Profile)
+	mux.HandleFunc("/debug/pprof/symbol", hpprof.Symbol)
+	mux.HandleFunc("/debug/pprof/trace", hpprof.Trace)
+	p := Profiler{
+		server: &http.Server{
+			Addr:    fmt.Sprintf(":%d", port),
+			Handler: mux,
+		},
+	}
+	return &p
+}
+
+// Run starts the HTTP pprof in the background.
+func (p *Profiler) Run() {
+	go func() {
+		p.mu.Lock()
+		defer p.mu.Unlock()
```

Comment:
This mutex is never released as `p.server.ListenAndServe()` is blocking.

Reply:
if you call shutdown, listen and server will exit, right?

> // When Shutdown is called, Serve, ListenAndServe, and
// ListenAndServeTLS immediately return ErrServerClosed. Make sure the
// program doesn't exit and waits instead for Shutdown to return.

Reply:
ah but actually Shutdown won't be able to grab a mutex

Reply:
why is this mutex necessary?

Reply:
That's a good question. `p.server` never changes after initialization so... is it unnecessary?

Reply:
It's not just a good question, it simply gets it: The mutex is unnecessary. IMHO I once added it because thinking of more even changing fields of the profiler. But those don't exist. Will remove it.